[WIP] Bugfix for "--topics/--services cannot work without extra option before bag file path"

MichaelOrlov opened the pull request:
- Closes https://github.com/ros2/rosbag2/issues/1628
- Add a quote guarded space-separated list type to player CLI
- For example, after the fix, the list of topics could be passed as:
1. `ros2 bag play --topics "topic1 topic2" bagpath`
2. `ros2 bag play --topics 'topic1 topic2' bagpath`
3. `ros2 bag play --topics topic1 bagpath`

morlov-apexai committed:
Add a quote guarded space separated list type to player CLI

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add unit tests for player args parser

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

fujitatomoya commented on ros2bag/ros2bag/verb/play.py:
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
+    parser.add_argument(
+        '-r', '--rate', type=check_positive_float, default=1.0,
+        help='rate at which to play back messages. Valid range > 0.0.')
+    parser.add_argument(
+        '--topics', type=lambda s: list(s.split(' ')), default=[], metavar='topic',
+        help='Space-delimited list of topics to play.')
```
- `ros2 bag play --topics chatter rosbag2_2024_04_25-10_55_16` works.
- `ros2 bag play --topics 'chatter greetings' rosbag2_2024_04_25-10_55_16` works.
- `ros2 bag play --topics "chatter greetings" rosbag2_2024_04_25-10_55_16` works.

but,

- `ros2 bag play --topics chatter greetings rosbag2_2024_04_25-10_55_16` fails with `ros2 bag play: error: argument bag_path: Bag path 'greetings' does not exist!`

i think we need to add, the list needs to be quoted.

```suggestion
    parser.add_argument(
        '--topics', type=lambda s: list(s.split(' ')), default=[], metavar='topic',
        help='Space-delimited and quoted list of topics to play.')
```

Barry-Xu-2018 commented on ros2bag/ros2bag/verb/play.py:
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
+        help='Space-delimited list of topics to play.')
+    parser.add_argument(
+        '--services', type=lambda s: list(s.split(' ')), default=[], metavar='service',
+        help='Space-delimited list of services to play.')
```
As fujitatomoya mentioned, "and quoted" should also be added to the description.

Barry-Xu-2018 commented on ros2bag/ros2bag/verb/play.py:
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
+             'Value must be positive. Defaults to not publishing.')
+    clock_args_group.add_argument(
+        '--clock-topics', type=lambda s: list(s.split(' ')), default=[],
+        help='List of topics separated by spaces that will trigger a /clock update '
```
As fujitatomoya mentioned, "and quoted" should also be added to the description.

MichaelOrlov commented:
@Barry-Xu-2018  @fujitatomoya Thank you for the preliminary review and opinion.   
I agree that we need to add in the parameters help section that the list needs to be surrounded by quotes.
I will follow up, fix it, and add more tests. I am also going to change lists for the recorder to be consistent.

Barry-Xu-2018 commented:
@MichaelOrlov 

>  I am also going to change lists for the recorder to be consistent.

If you have higher-priority tasks, I can take care of this modification.

morlov-apexai committed:
Add more unit tests for player args parser

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Update CLI args help section that lists shall be surrounded with quotes

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fix for failing integration tests which uses parameter lists

- Use lists surrounded with double quote marks in play_end_to_end test

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add space delimited quoted list for player "--remap" CLI argument

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add a quote guarded space separated list type to recorder CLI

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add unit tests for recorder args parser

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add a quote guarded space separated list type to burst CLI

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@Barry-Xu-2018 @fujitatomoya Almost done. 
I added changes for recorder and burst verbs to be consistent and fixed integration unit tests, which started failing after the changes.  

What I don't like in the current implementation is that in the recorder the `topics` CLI argument is positional.  
For instance, if one wants to specify a list of services and list of topics for recording the CLI command should be
```bash
ros2 bag record --services "service1 service2" "topic1 topic2" --output path/to/bag
```
It works currently, but the sequence of the parameters `--services "service1 service2" "topic1 topic2"` could cause confusion and could be a source of mistakes because the **topics** parameter can't be specified explicitly. i.e. `--topics`.
Please note that we inherited this behaviour from the past when we were implementing the services recording feature.

I am going to make the `topics` CLI argument as non-positional argument to be specified explicitly as `--topics`.  
Thoughts or concerns about it?

Barry-Xu-2018 commented:
@MichaelOrlov 

Yes, the argument for topics in the record command are styled differently from those in the play command. I considered making topics argument of the record command similar to the play command while I modified the code of record.  Considering compatibility, I ultimately decided not to propose this idea.  

I strongly agree with changing topics to a non-positional argument.

fujitatomoya commented:
> It works currently, but the sequence of the parameters --services "service1 service2" "topic1 topic2" could cause confusion and could be a source of mistakes because the topics parameter can't be specified explicitly. i.e. --topics.

agree, i do not think this is good behavior at all.

> I am going to make the topics CLI argument as non-positional argument to be specified explicitly as --topics.
> Thoughts or concerns about it?

sounds good to me.

MichaelOrlov commented:
- Closing as will not be merged in favor of the existent workaround with trailing `--`.
It turns out that the Python argparse class has a builtin option to split positional arguments by `--`
It is not well documented, but exists https://stackoverflow.com/questions/26985650/argparse-do-not-catch-positional-arguments-with-nargs/26986546#26986546

fujitatomoya commented:
@MichaelOrlov thanks, good to know.
